File: app/src/main/java/com/zionhuang/music/youtube/YouTubeAPIService.java
package com.zionhuang.music.youtube;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.preference.PreferenceManager;

import com.google.api.client.http.javanet.NetHttpTransport;
import com.google.api.client.json.gson.GsonFactory;
import com.google.api.services.youtube.YouTube;
import com.google.api.services.youtube.model.SearchListResponse;
import com.google.api.services.youtube.model.VideoListResponse;
import com.zionhuang.music.R;

import io.reactivex.rxjava3.core.Single;

public class YouTubeAPIService implements SharedPreferences.OnSharedPreferenceChangeListener {
    private static final String TAG = "YoutubeAPIService";
    private final String API_KEY_RID;
    private static String API_KEY;
    private static final String REGION_CODE = "US";
    private static final String CATEGORY_MUSIC = "10";
    private YouTube mYouTube;

    public YouTubeAPIService(Context context) {
        mYouTube = new YouTube.Builder(new NetHttpTransport.Builder().build(), new GsonFactory(), null)
                .setApplicationName(context.getResources().getString(R.string.app_name))
                .build();
        API_KEY_RID = context.getString(R.string.api_key);
        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
        API_KEY = sharedPreferences.getString(API_KEY_RID, "");
        sharedPreferences.registerOnSharedPreferenceChangeListener(this);
    }

    public Single<SearchListResponse> search(String query, String pageToken) {
        return Single.fromCallable(() ->
                mYouTube.search().list("snippet")
                        .setKey(API_KEY)
                        .setQ(query)
                        .setPageToken(pageToken)
                        .setMaxResults(20L)
                        .execute());
    }

    public Single<VideoListResponse> popularMusic(String pageToken) {
        return Single.fromCallable(() ->
                mYouTube.videos().list("snippet,contentDetails,statistics")
                        .setKey(API_KEY)
                        .setChart("mostPopular")
                        .setVideoCategoryId(CATEGORY_MUSIC)
                        .setRegionCode(REGION_CODE)
                        .setPageToken(pageToken)
                        .setMaxResults(20L)
                        .execute());
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (key.equals(API_KEY_RID)) {
            API_KEY = sharedPreferences.getString(key, "");
        }
    }
}

File: app/src/main/java/com/zionhuang/music/youtube/YouTubeAPIService.kt
package com.zionhuang.music.youtube

import android.content.Context
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.util.Log
import androidx.preference.PreferenceManager
import com.google.api.client.http.javanet.NetHttpTransport
import com.google.api.client.json.gson.GsonFactory
import com.google.api.services.youtube.YouTube
import com.google.api.services.youtube.model.SearchListResponse
import com.google.api.services.youtube.model.VideoListResponse
import com.zionhuang.music.R
import com.zionhuang.music.utils.PreferenceHelper
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class YouTubeAPIService(context: Context) {
    companion object {
        private const val TAG = "YoutubeAPIService"
        private const val REGION_CODE = "TW"
        private const val CATEGORY_MUSIC = "10"
    }

    private val apiKey = PreferenceHelper<String>(context, R.string.api_key)
    private val youTube: YouTube = YouTube.Builder(NetHttpTransport.Builder().build(), GsonFactory(), null)
            .setApplicationName(context.resources.getString(R.string.app_name))
            .build()

    @Suppress("BlockingMethodInNonBlockingContext")
    suspend fun search(query: String, pageToken: String?): SearchListResponse = withContext(Dispatchers.IO) {
        return@withContext youTube.search().list("snippet")
                .setKey(apiKey.value)
                .setQ(query)
                .setPageToken(pageToken)
                .setMaxResults(20L)
                .execute()
    }

    @Suppress("BlockingMethodInNonBlockingContext")
    suspend fun popularMusic(pageToken: String?): VideoListResponse = withContext(Dispatchers.IO) {
        youTube.videos().list("snippet,contentDetails,statistics")
                .setKey(apiKey.value)
                .setChart("mostPopular")
                .setVideoCategoryId(CATEGORY_MUSIC)
                .setRegionCode(REGION_CODE)
                .setPageToken(pageToken)
                .setMaxResults(20L)
                .execute()
    }
}